cli/auth.go

```diff
@@ -61,3 +67,94 @@ func (a *BasicAuth) OnRequest(req *http.Request, key string, params map[string]s
 	req.SetBasicAuth(params["username"], params["password"])
 	return nil
 }
+
+// ExternalToolAuth defers authentication to a third party tool.
+// This avoids baking all possible authentication implementations
+// inside restish itself.
+type ExternalToolAuth struct{}
+
+// Request is used to exchange requests with the external tool.
+type Request struct {
+	Method string      `json:"method"`
+	URI    string      `json:"uri"`
+	Header http.Header `json:"headers"`
+	Body   string      `json:"body"`
+}
+
+// Parameters defines the ExternalToolAuth parameter names.
+// A single parameter is supported and required: `commandline` which
+// points to the tool to call to authenticate a request.
+func (a *ExternalToolAuth) Parameters() []AuthParam {
+	return []AuthParam{
+		{Name: "commandline", Required: true},
+	}
+}
+
+// OnRequest gets run before the request goes out on the wire.
+// The supplied commandline argument is ran with a JSON input
+// and expects a JSON output on stdout
+func (a *ExternalToolAuth) OnRequest(req *http.Request, key string, params map[string]string) error {
+	commandLine, _ := params["commandline"]
+	shell, shellPresent := os.LookupEnv("SHELL")
+	if !shellPresent {
+		shell = "/bin/sh"
+	}
+	cmd := exec.Command(shell, "-c", commandLine)
+	stdin, err := cmd.StdinPipe()
+	if err != nil {
+		return err
+	}
+
+	bodyStr := ""
+	if req.Body != nil {
+		bodyBytes, err := io.ReadAll(req.Body)
+		if err != nil {
+			return err
+		}
+		bodyStr = string(bodyBytes)
+		req.Body = io.NopCloser(strings.NewReader(bodyStr))
+	}
+
+	textRequest := Request{
+		Method: req.Method,
+		URI:    req.URL.String(),
+		Header: req.Header,
+		Body:   bodyStr,
+	}
+	requestBytes, err := json.Marshal(textRequest)
+	if err != nil {
+		return err
+	}
+	_, err = stdin.Write(requestBytes)
+	if err != nil {
+		return err
+	}
+	stdin.Close()
+	outBytes, err := cmd.Output()
+	if err != nil {
+		return err
+	}
+	if len(outBytes) <= 0 {
+		return nil
+	}
+	var requestUpdates Request
+	err = json.Unmarshal(outBytes, &requestUpdates)
+	if err != nil {
+		return err
+	}
```

Comment:
I'm not super familiar with this, but do you think there are potential buffer problems with very large requests? I see the examples (e.g. [this one](https://pkg.go.dev/os/exec#Cmd.StdinPipe)) using streaming decoders or goroutines and a `cmd.Wait()` to handle this sort of thing. May need to dig into what gotchas there are here before merging.

Reply:
Hi @danielgtaylor, thanks for the review. The current assumes that for most APIs inputs have somewhat moderate sizes thus optimizes for this.

Most commonly available APIs (excluding S3 which is the odd one out, but doesn't expose OpenAPI so less relevant here) which allow large body inputs do not include the body in their signature mechanisms. 

If the above seems sensible to you, I can adapt the patch to allow controlling what the auth mechanism hands over to the sidecar script (allowing to exclude the body when unnecessary).

Reply:
I adapted as discussed, adding an `omitbody` parameter which when preset does not buffer the input body in memory and forwards it down untouched